opentelemetry/proto/metrics/v1/metrics.proto

```diff
@@ -87,23 +87,50 @@ message InstrumentationLibraryMetrics {
 //                       +---------------------------+
 //
 //-----------------------------------------------------------------------
-// DataPoint is a value of specific type corresponding to a given moment in
-// time. Each DataPoint is timestamped.
-//
-// DataPoint is strongly typed: each DataPoint type has a specific Protobuf message
-// depending on the value type of the metric and thus there are currently 4 DataPoint
-// messages, which correspond to the types of metric values.
 message Metric {
   // metric_descriptor describes the Metric.
   MetricDescriptor metric_descriptor = 1;
 
-  // Data is a list of one or more DataPoints for a single metric. Only one of the
-  // following fields is used for the data, depending on the type of the metric defined
-  // by MetricDescriptor.type field.
-  repeated Int64DataPoint int64_data_points = 2;
-  repeated DoubleDataPoint double_data_points = 3;
-  repeated HistogramDataPoint histogram_data_points = 4;
-  repeated SummaryDataPoint summary_data_points = 5;
+  repeated Data data = 2;
+
+  message Data {
+    // The set of labels that uniquely identify this timeseries.
+    //
+    // These labels take precedence over any labels with the same key defined
+    // in the metric descriptor.
+    repeated opentelemetry.proto.common.v1.StringKeyValue labels = 1;
+
+    // start_time_unix_nano is the beginning of a time interval over which
+    // this data point was recorded. It MUST be represented as the UNIX Epoch
+    // time in nanoseconds (nanoseconds since 00:00:00 UTC on 1 January 1970).
+    // If zero, it is treated as unspecified and may be set by the receiver.
+    //
+    // If this data point is a part of a timeseries for an INSTANTANEOUS
+    // metric, this value is ignored.
+    //
+    // If this data point is a part of a timeseries for an DELTA, or
+    // CUMULATIVE metric, this value represents the instant before the
+    // measuring time interval began (i.e. for time t to be in the measuring
+    // interval than start_time_unix_nano < t <= time_unix_nano).
+    //
+    // This time MUST be reset so as to represent a new measurement lifetime
+    // if this data point is part of a timeseries for a CUMULATIVE metric and
+    // the metric value is reset to zero.
+    fixed64 start_time_unix_nano = 2;
```

Comment:
I've read and re-read this explanation, and I think you're saying what I think we want, but we should clarify this, especially for cumulative values. I found a helpful explanation in Stackdriver's documentation:

https://cloud.google.com/monitoring/api/ref_v3/rest/v3/TimeSeries#point

> The time interval to which the data point applies. For GAUGE metrics, the start time is optional, but if it is supplied, it must equal the end time. For DELTA metrics, the start and end time should specify a non-zero interval, with subsequent points specifying contiguous and non-overlapping intervals. For CUMULATIVE metrics, the start and end time should specify a non-zero interval, with subsequent points specifying the same start time and increasing end times, until an event resets the cumulative value to zero and sets a new start time for the following points.

For DELTA instruments: can we recommend that start and end times equal the start and end time of the interval? If so, in a simple configuration where all metrics are collected on the same interval, all Instantaneous and Delta measurements can infer their end (Instantaneous) or start/end (Delta) time from the report and be skipped in the Data field.  At least I think we should have an option to leave out timestamps when they can be inferred from the report.

We have to pay particular attention to how cumulatives are reported.  The start-time of a cumulative should remain constant as long as the cumulative is not reset, as stated in the Stackdriver docs. It means that the SDK will be responsible for remembering cumulative values that are actively being reported.  When some number of intervals pass without reporting that cumulative, the SDK may forget it, as follows.

1. Remember the end timestamp fo the last interval that reported the metric
2. Reset **all** values for the instrument, retain only the memory of the end timestamp
3. Any new cumulative value for the instrument (i.e., distinct label set) will be reported as having a start time equal to the last timestamp where the instrument was reset.

Item (2) ensures that we can aggregate cumulative values for a single metric within the process, since all values will have the same reset time.  This does not ensure that multiple instruments can be aggregated in the process: an error condition would arise as we try to aggregate multiple cumulative values with different star times.  The same condition may arise in a distributed setting, where if multiple processes report the same metric with different cumulative start times, aggregation is not possible.  I suggest we leave it at that, there is a known condition where cumulatives cannot be aggregated except when they have the same start time.

Reply:
Thanks for the suggestions, they are much appreciated.

I've updated the time interval handling. PTAL and let me know if you have any other suggestions.